addons/account_payment/controllers/payment.py

```diff
@@ -17,11 +17,12 @@ def invoice_pay_form(self, acquirer_id, invoice_id, save_token=False, access_tok
         :return html: form containing all values related to the acquirer to
                       redirect customers to the acquirer website """
         success_url = kwargs.get('success_url', '/my')
-        callback_method = kwargs.get('callback_method', '')
-
         invoice_sudo = request.env['account.invoice'].sudo().browse(invoice_id)
         if not invoice_sudo:
             return False
+        # Check if the current user has access to this invoice
+        if invoice_sudo.partner_id.id not in [request.env.user.partner_id.id, request.env.user.partner_id.commercial_partner_id.id]:
```

Comment:
The invoice itself has a `commercial_partner_id` field, and we should compare these so that the hierarchies are properly compared:

```python
req_partner = request.env.user.partner_id
if invoice_sudo.commercial_partner_id != req_partner.commercial_partner_id:
    ...
```